### PR TITLE
ARROW-12591: [Java][Gandiva] Create single Gandiva jar for MacOS and Linux

### PR DESCRIPTION
**Pull request by anthonylouisbsb**
Today, there are two different tasks that generate Gandiva's jars for MacOS and Linux. The objective is to create a single jar with the shared lib for the two operating systems.

### CODE REVIEW

**Comment by projjal** [comment]
@github-actions crossbow submit -g gandiva-jar

**Comment by projjal** [comment]
@github-actions crossbow submit -g gandiva

**Review by projjal — APPROVED** [review]

**Review by kou — COMMENTED** [review]
Thanks!

I hope that we also build all other Java packages building in our release process ( https://github.com/apache/arrow/blob/master/dev/release/post-11-java.sh ) like this as the next task. We can just upload all Java packages including the Gandiva Java package in our release process by the next task.

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    GANDIVA_LIB="$CPP_BUILD_DIR"libgandiva_jni.so
+    WHITELIST=(linux-vdso libz librt libdl libpthread libstdc++ libm libgcc_s libc ld-linux-x86-64)
+else
+    SO_DEP="otool -L"
+    GANDIVA_LIB="$CPP_BUILD_DIR"libgandiva_jni.dylib
+    WHITELIST=(libgandiva_jni libz libncurses libSystem libc++)
+fi
+
+# print the shared library dependencies
+eval "$SO_DEP" "$GANDIVA_LIB"
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: Changes applied

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        found=false
+        for item in "${WHITELIST[@]}"
+        do
+        if [[ "$line" == *"$item"* ]] ; then
+            found=true
+        fi
+        done
+        if [[ "$found" == false ]] ; then
+        echo "Unexpected shared dependency found"
+        exit 1
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: changes applied

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        fi
+    done < <(eval "$SO_DEP" "$GANDIVA_LIB" | awk '{print $1}')
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: It is a good idea. Suggestion applied

File: dev/tasks/gandiva-jars/github.linux.macos.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          set -e
+          docker run -v $PWD:/arrow quay.io/anthonylouisbsb/arrow:gandivadocker /arrow/dev/tasks/gandiva-jars/build-cpp-linux.sh
+          dev/tasks/gandiva-jars/check_shared_dependencies.sh
+        env:
+          OS_NAME: "linux"
+          CHECK_SHARED_DEPENDENCIES: true
+      - name: Upload Artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: ubuntu-shared-lib-{{ '${{ github.sha }}' }}
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/github.linux.macos.yml]: I put just to generate a unique artifact name, but I think it is not necessary, I removed it.

File: dev/tasks/gandiva-jars/github.linux.macos.yml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          path: {{ '${{ github.workspace }}' }}/arrow/dist/libgandiva_jni.so
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/github.linux.macos.yml]: Change applied

File: dev/tasks/gandiva-jars/github.linux.macos.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          dev/tasks/gandiva-jars/check_shared_dependencies.sh
+        env:
+          OS_NAME: "osx"
+          CHECK_SHARED_DEPENDENCIES: true
+          MACOSX_DEPLOYMENT_TARGET: "10.11"
+      - name: Upload Artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: macos-shared-lib-{{ '${{ github.sha }}' }}
+          path: {{ '${{ github.workspace }}' }}/arrow/dist/libgandiva_jni.dylib
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/github.linux.macos.yml]: change applied

File: dev/tasks/gandiva-jars/github.linux.macos.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+          git clone --no-checkout {{ arrow.remote }} arrow
+          git -C arrow fetch -t {{ arrow.remote }} {{ arrow.branch }}
+          if [ $CROSSBOW_USE_COMMIT_ID = true ]; then git -C arrow checkout {{ arrow.head }}; else git -C arrow checkout FETCH_HEAD; fi
+          git -C arrow submodule update --init --recursive
+          mkdir -p arrow/dist
+      - name: Download Linux Gandiva Library
+        uses: actions/download-artifact@v2
+        with:
+          name: ubuntu-shared-lib-{{ '${{ github.sha }}' }}
+          path: {{ '${{ github.workspace }}' }}/arrow/dist
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/github.linux.macos.yml]: change applied

File: dev/tasks/gandiva-jars/github.linux.macos.yml [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      - name: Download MacOS Gandiva Library
+        uses: actions/download-artifact@v2
+        with:
+          name: macos-shared-lib-{{ '${{ github.sha }}' }}
+          path: {{ '${{ github.workspace }}' }}/arrow/dist
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/github.linux.macos.yml]: change applied

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -e
+
+CPP_BUILD_DIR=$GITHUB_WORKSPACE/arrow/dist/
+
+if [[ $OS_NAME == "linux" ]]; then
+    SO_DEP=ldd
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: I changed the indentation for the file to use 2-spaces.

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh [hunk 10]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: The suggestion was applied

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: sugestion applied

File: dev/tasks/gandiva-jars/check_shared_dependencies.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (anthonylouisbsb) [reply dev/tasks/gandiva-jars/check_shared_dependencies.sh]: suggestion applied

**Comment by anthonylouisbsb** [comment]
@kou All suggestions were applied.

**Comment by kou** [comment]
@github-actions crossbow submit gandiva-jar

**Review by kou — APPROVED** [review]
+1

**Comment by anthonylouisbsb** [comment]
> Thanks!
> 
> I hope that we also build all other Java packages building in our release process ( https://github.com/apache/arrow/blob/master/dev/release/post-11-java.sh ) like this as the next task. We can just upload all Java packages including the Gandiva Java package in our release process by the next task.

@kou I am interested in executing that task. Is there some opened JIRA ticket for it? Or should I create a new one?

**Comment by kou** [comment]
Great!

Could you create a new JIRA issue for it?
The new JIRA issue is a child of ARROW-11135 like ARROW-12591.

In the task, we need to do the followings:

1. Rename the `gandiva-jar` task to `java-jar`
2. Build all Java packages in the task

We'll update https://github.com/apache/arrow/blob/master/dev/release/post-11-java.sh and related release related scripts after the task.

**Comment by anthonylouisbsb** [comment]
@kou I created the JIRA issue https://issues.apache.org/jira/browse/ARROW-12699.

I will start to work on it.

**Pull request closed without merging**